control-plane/csi-controller/src/client.rs

```diff
@@ -0,0 +1,373 @@
+macro_rules! collection_getter {
+    ($name:ident, $t:ty, $urn:expr) => {
+        pub async fn $name(&self) -> Result<Vec<$t>, ApiClientError> {
+            self.get_collection::<$t>($urn).await
+        }
+    };
+}
+
+impl MayastorApiClient {
+    async fn get_collection_item<R>(&self, urn: UrnType<'_>) -> Result<R, ApiClientError>
+    where
+        for<'a> R: Deserialize<'a>,
+    {
+        let response = self.do_get(&urn).await.map_err(|e| {
+            ApiClientError::ServerCommunicationError(format!(
+                "Failed to get {:?}, error = {}",
+                urn, e
+            ))
+        })?;
+
+        // Check HTTP status code.
+        match response.status() {
+            StatusCode::OK => {}
+            StatusCode::NOT_FOUND => {
+                let (rtype, rname) = urn.classify();
+                return Err(ApiClientError::ResourceNotExists(format!(
+                    "{} {} not found",
+                    rtype, rname
+                )));
+            }
+            http_status => {
+                return Err(ApiClientError::GenericOperationError(format!(
+                    "Failed to GET {:?}, HTTP error = {}",
+                    urn, http_status,
+                )))
+            }
+        };
+
+        // Get response body if request succeeded.
+        let body = response.bytes().await.map_err(|e| {
+            ApiClientError::InvalidResponseError(format!(
+                "Failed to obtain body from HTTP response while getting {}, error = {}",
+                urn, e,
+            ))
+        })?;
+
+        serde_json::from_slice::<R>(&body).map_err(|e| {
+            ApiClientError::InvalidResponseError(format!(
+                "Failed to deserialize object {}, error = {}",
+                std::any::type_name::<R>(),
+                e
+            ))
+        })
+    }
+
+    // Get one resource instance.
+    async fn do_get(&self, urn: &UrnType<'_>) -> Result<Response, Error> {
+        let u = format!("{}/{}", self.base_url, urn);
+        let uri = Url::parse(&u).unwrap();
+
+        self.rest_client.get(uri).send().await
+    }
+
+    // Perform resource deletion, optionally idempotent.
+    async fn do_delete(&self, urn: &UrnType<'_>, idempotent: bool) -> Result<(), ApiClientError> {
+        let u = format!("{}/{}", self.base_url, urn);
+        let uri = Url::parse(&u).unwrap();
+
+        let response = self.rest_client.delete(uri).send().await.map_err(|e| {
+            ApiClientError::ServerCommunicationError(format!(
+                "DELETE {} request failed, error={}",
+                u, e
+            ))
+        })?;
+
+        // Check HTTP status code, handle DELETE idempotency transparently.
+        let res = match response.status() {
+            StatusCode::OK => Ok(()),
+            // Handle idempotency as requested by the caller.
+            StatusCode::NOT_FOUND | StatusCode::NO_CONTENT | StatusCode::PRECONDITION_FAILED => {
+                if idempotent {
```

Comment:
The idea is that we handle operation as successful even if there are no resource exists (volume removal or so). In case of missing HTTP resource we will see 404, which is an error from HTTP server perspective, but not an error for us in case of volume removal, as per CSI spec.